airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/load/component/TableSchemaEvolutionSuite.kt

```diff
@@ -0,0 +1,469 @@
+                ),
+            )
+        val modifiedSchema =
+            ObjectType(
+                linkedMapOf(
+                    "to_retain" to FieldType(StringType, true),
+                    "to_change" to FieldType(StringType, true),
+                    "to_add" to FieldType(StringType, true),
+                ),
+            )
+        val modifiedStream =
+            Fixtures.createAppendStream(
+                namespace = testTable.namespace,
+                name = testTable.name,
+                schema = modifiedSchema,
+            )
+
+        // Create the table and compute the schema changeset
+        val (_, expectedSchema, changeset) =
+            computeSchemaEvolution(
+                testTable,
+                initialSchema,
+                initialColumnNameMapping,
+                modifiedSchema,
+                modifiedColumnNameMapping,
+            )
+        // Insert a record before applying the changeset
+        testClient.insertRecords(
+            testTable,
+            initialColumnNameMapping,
+            mapOf(
+                COLUMN_NAME_AB_RAW_ID to StringValue("fcc784dd-bf06-468e-ad59-666d5aaceae8"),
+                COLUMN_NAME_AB_EXTRACTED_AT to TimestampWithTimezoneValue("2025-01-22T00:00:00Z"),
+                COLUMN_NAME_AB_META to ObjectValue(linkedMapOf()),
+                COLUMN_NAME_AB_GENERATION_ID to IntegerValue(1),
+                "to_retain" to StringValue("to_retain original value"),
+                "to_change" to IntegerValue(42),
+                "to_drop" to StringValue("to_drop original value"),
+            ),
+        )
+
+        client.applyChangeset(
+            modifiedStream,
+            modifiedColumnNameMapping,
+            testTable,
+            expectedSchema.columns,
+            changeset,
+        )
+
+        val postAlterationRecords = harness.readTableWithoutMetaColumns(testTable)
+        Assertions.assertEquals(
+            listOf(
+                mapOf(
+                    "to_retain" to "to_retain original value",
+                    // changed from int to string
+                    "to_change" to "42",
+                    // note the lack of `to_add` - new columns should be initialized to null
+                    )
+            ),
+            postAlterationRecords
+                .removeNulls()
+                .reverseColumnNameMapping(modifiedColumnNameMapping, airbyteMetaColumnMapping),
+        ) {
+            "Expected records were not in the overwritten table."
+        }
+
+        val postAlterationDiscoveredSchema = client.discoverSchema(testTable)
+        val postAlterationChangeset =
+            client.computeChangeset(postAlterationDiscoveredSchema.columns, expectedSchema.columns)
+        assertTrue(
+            postAlterationChangeset.isNoop(),
+            "After applying the changeset, we should be a noop against the expected schema"
+        )
+    }
+
```

Comment:
Basic tests look good to me. I think we might want some tests that can be Parameterized over an arbitrary list of inputs and outputs. Can be added later ofc.

Reply:
yeah, I think all the pieces are mostly there already. like the `discover` / `computeSchema` tests are basically that, and connector implementers could do `@ParameterizedTest` if they wanted to

maybe worth trying to make a nice abstraction for the applyChangeset piece at some point (or maybe the full e2e test), but IMO that's less pressing than the other stuff we have planned. I.e. it's a thing that we know is possible, would just take a little time to design+type up.